public/app/features/logs/UniqueKeyMaker.ts

```diff
@@ -0,0 +1,25 @@
+// this class generates react-keys that are guaranteed to be unique.
+// it will try to use the provided `maybeId`, but if that's a duplicate,
+// it will use an index-based key.
+// NOTE: it will always add a prefix to the string, this is necessary
+// to avoid the problem if your proposed key would conflict with
+// the index-based names
+export class UniqueKeyMaker {
+  seen: Set<string>;
+  count: number;
+  constructor() {
+    this.seen = new Set();
+    this.count = 0;
+  }
+
+  getKey(proposedKey: string) {
+    this.count += 1;
+    const maybeKey = `k_${proposedKey}`;
+    if (this.seen.has(maybeKey)) {
+      return `i_${this.count}`;
+    } else {
+      this.seen.add(maybeKey);
+      return maybeKey;
+    }
+  }
```

Comment:
At this point this feels nitpicky, but maybe you'd agree. I would only override the key if it's seen, otherwise I'll return it unmodified.

```suggestion
  getKey(proposedKey: string) {
    this.count += 1;
    if (this.seen.has(proposedKey)) {
      // I would also consider 1) keeping the original key and just prefixing it and 2) pass the new key through the `seen` check to be 100% sure it's not duplicated.
      return this.getKey(`i_${this.count}_${proposedKey}`);
    } else {
      this.seen.add(proposedKey);
      return proposedKey;
    }
  }
```

Reply:
i see your point, but the problem is that that does not guarantee 100% uniqueness..
imagine these keys:
- one
- i_3
- one

currently this produces the following keys:
- k_one
- k_i_3
- i_3


if we don't prefix unique keys, it will produce this:
- one
- i_3
- i_3

which is bad.

technically we could also `seen`-check the index-based values, but then there's a problem. when you seen-check an indexed-value, and see that it has been used, then what? increment the index and try again? what if that one is used too? again increment? how many times? etc.

i don't think it's unsolvable, but right now i don't have any ideas that would be simple and do a 100% guarantee... but i'm very open to ideas.  i will merge this one, i hope that's ok, and we can improve it if we come up with a good idea 👍

Reply:
Makes sense! Thanks for checking.